ompi/mpiext/pcollreq/c/neighbor_alltoallw_init.c

```diff
@@ -151,7 +151,9 @@ int MPIX_Neighbor_alltoallw_init(const void *sendbuf, const int sendcounts[], co
                                                      info, request,
                                                      comm->c_coll->coll_neighbor_alltoallw_init_module);
     if (OPAL_LIKELY(OMPI_SUCCESS == err)) {
-        ompi_coll_base_retain_datatypes_w(*request, sendtypes, recvtypes);
+        ompi_coll_base_retain_datatypes_w(*request,
+                                          (MPI_Datatype*) sendtypes,
```

Comment:
The need for this is unclear to me. The sendtypes is an "const MPI_Datatype []" and the function ompi_coll_base_retain_datatypes_w expects "ompi_datatype_t * const []". Why passing through an "MPI_Datatype*"(aka ompi_datatype_t*), which drop a const, is a more acceptable answer than just passing the sendtypes with their original type ?

Reply:
My understanding is that `ompi_datatype_t * const X[]` marks `X` `const`, i.e., the pointer value cannot change but the value(s) pointed are still mutable.

Reply:
yes, it's about a pointer to const versus a const pointer. But it is strange that going through a cast without const will solve the warning.

Reply:
Without the cast we're passing a non-const pointer to const data (`const MPI_Datatype sendtypes[]`) to a const pointer to mutable data, so we need to cast away const for the data. The `const` in the `ompi_coll_base_retain_datatypes_w` declaration really shouldn't be there because it is irrelevant for the interface and only relevant for the implementation (telling the compiler to disallow changing the pointer value in `stypes` within `ompi_coll_base_retain_datatypes_w`). It does lead to confusion though...

Reply:
Or that the MPI API is incorrect when it assumes the MPI objects should not change, instead of assuming simply that the pointers in the array should not change.

Reply:
Yes, the correct function signature should be:

```
int ompi_coll_base_retain_datatypes_w( ompi_request_t *request,
                                       ompi_datatype_t * const*stypes,
                                       ompi_datatype_t * const*rtypes);
```

Reply:
At the user level, the type is `const MPI_Datatype styles[]` (`const ompi_datatype_t **stypes` for Open MPI), so casting it to an `ompi_datatype_t * const *stypes` will drop the `const` modifier somehow.

So the `const` qualifier is a fit for MPICH (where a `MPI_Datatype` is an `int`) but not for Open MPI (where `MPI_Datatype` is a pointer).

So I am with @bosilca here, the MPI API is incorrect in the sense it only works if the `MPI_Datatype` opaque type is an `int`.

Reply:
I have to disagree here: the MPI interface is correct in that OMPI is not allowed to change any of the pointers in the `stypes` array (e.g., replace an `MPI_INT` with an `MPI_DOUBLE`). That is what the `const` in the MPI interface says. OMPI is still free to modify the objects pointed to by the elements in the `stypes` array, with the proper function signature for `ompi_coll_base_retain_datatypes_w` no cast is required (see my [previous comment](https://github.com/open-mpi/ompi/pull/8203#discussion_r522465838))

Reply:
let me take a second look at it. `const` is an issue if a datatype is free'd at the end, but i thought that was not the only one ... checking from now.

Reply:
@devreal I stand corrected and my previous analysis was simply wrong.

I just issued #8207 to (hopefully) correctly address this issue.

Reply:
Based on the discussion here and on #8207, I'll remove this (and similar) changes from this PR.

Reply:
Just curious: was the compiler warning resolved in https://github.com/open-mpi/ompi/pull/8207? I don't see anything changing the signature of `ompi_coll_base_retain_datatypes_w` there to avoid the original warning you tried to fix.